Python/Product/PythonTools/PythonTools/Editor/BraceCompletion/BraceCompletionContextProvider.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using Microsoft.VisualStudio.Text.BraceCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Utilities;
+using System.Diagnostics;
+
+namespace Microsoft.PythonTools.Editor.BraceCompletion {
+
+    [Export(typeof(IBraceCompletionContextProvider))]
+    [BracePair('(', ')')]
+    [BracePair('[', ']')]
+    [BracePair('{', '}')]
+    [BracePair('"', '"')]
+    [BracePair('\'', '\'')]
+    [ContentType(PythonCoreConstants.ContentType)]
+    internal sealed class BraceCompletionContextProvider : IBraceCompletionContextProvider {
+        public bool TryCreateContext(ITextView textView, SnapshotPoint openingPoint, char openingBrace, char closingBrace, out IBraceCompletionContext context) {
+            if (IsValidBraceCompletionContext(openingPoint)) {
+                context = new BraceCompletionContext();
+                return true;
+            } else {
+                context = null;
+                return false;
+            }
+        }
+
+        private bool IsValidBraceCompletionContext(SnapshotPoint openingPoint) {
+            Debug.Assert(openingPoint.Position >= 0, "SnapshotPoint.Position should always be zero or positive.");
+
+            if (openingPoint.Position > 0) {
+                var classificationSpans = openingPoint.Snapshot.TextBuffer.GetPythonClassifier().GetClassificationSpans(new SnapshotSpan(openingPoint - 1, 1));
+                foreach (var span in classificationSpans) {
```

Comment:
Afraid not - the brace completion service doesn't even try and trigger. Here's how the sequence of pressing `'` repeatedly looks:

1. Adds first `'` and provisional `'`
2. Overwrites provisional `'`, so now we have `''`
3. Adds third `'` and no brace completion request is made - we have `'''`
4. Repeat step 3 for all subsequent quotes

So when typing `'''`, we don't complete the closing three (we _could_ support that manually through the edit filter, but brace completion works on `char` and not `string`, so I don't think we'd get overwrite support easily), but it's also not especially painful.